x/clp/keeper/calculations.go

```diff
@@ -325,6 +325,51 @@ func calcPmtpFactor(r sdk.Dec) big.Rat {
 	return *one
 }
 
+func CalcSpotPriceNative(pool *types.Pool, decimalsExternal uint8, pmtpCurrentRunningRate sdk.Dec) (sdk.Dec, error) {
+	return CalcSpotPriceX(pool.NativeAssetBalance, pool.ExternalAssetBalance, types.NativeAssetDecimals, decimalsExternal, pmtpCurrentRunningRate, true)
+}
+
+func CalculateSpotPriceExternal(pool *types.Pool, decimalsExternal uint8, pmtpCurrentRunningRate sdk.Dec) (sdk.Dec, error) {
+	return CalcSpotPriceX(pool.ExternalAssetBalance, pool.NativeAssetBalance, decimalsExternal, types.NativeAssetDecimals, pmtpCurrentRunningRate, false)
+}
+
+// Calculates the spot price of asset X in the preferred denominations accounting for PMTP.
+// Since this method applies PMTP adjustment, one of X, Y must be the native asset.
+func CalcSpotPriceX(X, Y sdk.Uint, decimalsX, decimalsY uint8, pmtpCurrentRunningRate sdk.Dec, isXNative bool) (sdk.Dec, error) {
```

Comment:
Maybe also add a comment that this calculation is only used to compute values for the frontend and not for the chain itself?

Reply:
I've added tests for the more specialised ones. I've kept this function unhidden for now so that it can also be tested. We could add a comment to the PolicyRun method once we decide where this calculation should belong?